snapshots/overlay/overlay.go

```diff
@@ -292,6 +327,19 @@ func (o *snapshotter) Walk(ctx context.Context, fn snapshots.WalkFunc, fs ...str
 		return err
 	}
 	defer t.Rollback()
+	if o.upperdirLabel {
+		fn = func(ctx context.Context, info snapshots.Info) error {
+			id, _, _, err := storage.GetInfo(ctx, info.Name)
+			if err != nil {
+				return err
+			}
+			if info.Labels == nil {
+				info.Labels = make(map[string]string)
+			}
+			info.Labels[upperdirKey] = o.upperPath(id)
+			return fn(ctx, info)
```

Comment:
This looks like it would recurse since `fn` will get evaluated on each call

Reply:
https://play.golang.org/p/4AsEnaj_-37

Reply:
Thank you for catching this. Fixed.